node/forest_libp2p/bitswap/tests/go_compat.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+#[cfg(test)]
+mod tests {
+    use anyhow::Result;
+    use forest_libp2p_bitswap::{
+        BitswapBehaviour, BitswapBehaviourEvent, BitswapMessage, BitswapRequest, BitswapResponse,
+    };
+    use libipld::{
+        multihash::{self, MultihashDigest},
+        Cid,
+    };
+    use libp2p::{
+        core, futures::StreamExt, identity, noise, request_response::RequestResponseMessage,
+        swarm::SwarmEvent, tcp, yamux, PeerId, Swarm, Transport,
+    };
+    use std::{process::Command, time::Duration};
+
+    const TIMEOUT: Duration = Duration::from_secs(60);
+    const LISTEN_ADDR: &str = "/ip4/127.0.0.1/tcp/0";
+    const GO_APP_DIR: &str = "tests/go-app";
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn bitswap_go_compat_test() {
+        bitswap_go_compat_test_impl().await.unwrap()
+    }
+
+    async fn bitswap_go_compat_test_impl() -> Result<()> {
+        let id_keys = identity::Keypair::generate_ed25519();
+        let peer_id = PeerId::from(id_keys.public());
+        let transport = tcp::tokio::Transport::default()
+            .upgrade(core::upgrade::Version::V1)
+            .authenticate(noise::NoiseAuthenticated::xx(&id_keys).unwrap())
+            .multiplex(yamux::YamuxConfig::default())
+            .timeout(TIMEOUT)
+            .boxed();
+        let behaviour = BitswapBehaviour::new(&[b"/test/ipfs/bitswap/1.2.0"], Default::default());
+        let mut swarm = Swarm::with_tokio_executor(transport, behaviour, peer_id);
+        swarm.listen_on(LISTEN_ADDR.parse()?)?;
+        let expected_inbound_request_cid_str = "bitswap_request_from_go";
+        let expected_inbound_request_cid = Cid::new_v0(
+            multihash::Code::Sha2_256.digest(expected_inbound_request_cid_str.as_bytes()),
+        )?;
+        let outbound_request_cid =
+            Cid::new_v0(multihash::Code::Sha2_256.digest(b"bitswap_request_from_rust"))?;
+        let (local_addr_tx, local_addr_rx) = flume::unbounded();
+        let (inbound_request_tx, inbound_request_rx) = flume::unbounded();
+        let (inbound_response_tx, inbound_response_rx) = flume::unbounded();
+        tokio::spawn(async move {
```

Comment:
Let's refactor this method. I understand these are tests, but at the same time, this is the most prominent example of usage of this bitswap crate, so it should be as clean as possible. Folks will most likely copy-paste it.

Reply:
Thanks for elaborating, you convinced me.